File: src/java/main/org/apache/zookeeper/server/quorum/LeaderZooKeeperServer.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.zookeeper.server.quorum;

import java.io.IOException;

import org.apache.zookeeper.KeeperException.SessionExpiredException;
import org.apache.zookeeper.jmx.MBeanRegistry;
import org.apache.zookeeper.server.DataTreeBean;
import org.apache.zookeeper.server.FinalRequestProcessor;
import org.apache.zookeeper.server.PrepRequestProcessor;
import org.apache.zookeeper.server.RequestProcessor;
import org.apache.zookeeper.server.ServerCnxn;
import org.apache.zookeeper.server.SessionTrackerImpl;
import org.apache.zookeeper.server.ZKDatabase;
import org.apache.zookeeper.server.persistence.FileTxnSnapLog;

/**
 * 
 * Just like the standard ZooKeeperServer. We just replace the request
 * processors: PrepRequestProcessor -> ProposalRequestProcessor ->
 * CommitProcessor -> Leader.ToBeAppliedRequestProcessor ->
 * FinalRequestProcessor
 */
public class LeaderZooKeeperServer extends QuorumZooKeeperServer {
    CommitProcessor commitProcessor;

    /**
     * @param port
     * @param dataDir
     * @throws IOException
     */
    LeaderZooKeeperServer(FileTxnSnapLog logFactory, QuorumPeer self,
            DataTreeBuilder treeBuilder, ZKDatabase zkDb) throws IOException {
        super(logFactory, self.tickTime, self.minSessionTimeout,
                self.maxSessionTimeout, treeBuilder, zkDb, self);
    }

    public Leader getLeader(){
        return self.leader;
    }
    
    @Override
    protected void setupRequestProcessors() {
        RequestProcessor finalProcessor = new FinalRequestProcessor(this);
        RequestProcessor toBeAppliedProcessor = new Leader.ToBeAppliedRequestProcessor(
                finalProcessor, getLeader().toBeApplied);
        commitProcessor = new CommitProcessor(toBeAppliedProcessor,
                Long.toString(getServerId()), false);
        commitProcessor.start();
        ProposalRequestProcessor proposalProcessor = new ProposalRequestProcessor(this,
                commitProcessor);
        proposalProcessor.initialize();
        firstProcessor = new PrepRequestProcessor(this, proposalProcessor);
        ((PrepRequestProcessor)firstProcessor).start();
    }

    @Override
    public int getGlobalOutstandingLimit() {
        return super.getGlobalOutstandingLimit() / (self.getQuorumSize() - 1);
    }
    
    @Override
    protected void createSessionTracker() {
        sessionTracker = new SessionTrackerImpl(this, getZKDatabase().getSessionWithTimeOuts(),
                tickTime, self.getId());
        ((SessionTrackerImpl)sessionTracker).start();
    }


    public boolean touch(long sess, int to) {
        return sessionTracker.touchSession(sess, to);
    }

    @Override
    protected void registerJMX() {
        // register with JMX
        try {
            jmxDataTreeBean = new DataTreeBean(getZKDatabase().getDataTree());
            MBeanRegistry.getInstance().register(jmxDataTreeBean, jmxServerBean);
        } catch (Exception e) {
            LOG.warn("Failed to register with JMX", e);
            jmxDataTreeBean = null;
        }
    }

    public void registerJMX(LeaderBean leaderBean,
            LocalPeerBean localPeerBean)
    {
        // register with JMX
        if (self.jmxLeaderElectionBean != null) {
            try {
                MBeanRegistry.getInstance().unregister(self.jmxLeaderElectionBean);
            } catch (Exception e) {
                LOG.warn("Failed to register with JMX", e);
            }
            self.jmxLeaderElectionBean = null;
        }

        try {
            jmxServerBean = leaderBean;
            MBeanRegistry.getInstance().register(leaderBean, localPeerBean);
        } catch (Exception e) {
            LOG.warn("Failed to register with JMX", e);
            jmxServerBean = null;
        }
    }

    @Override
    protected void unregisterJMX() {
        // unregister from JMX
        try {
            if (jmxDataTreeBean != null) {
                MBeanRegistry.getInstance().unregister(jmxDataTreeBean);
            }
        } catch (Exception e) {
            LOG.warn("Failed to unregister with JMX", e);
        }
        jmxDataTreeBean = null;
    }

    protected void unregisterJMX(Leader leader) {
        // unregister from JMX
        try {
            if (jmxServerBean != null) {
                MBeanRegistry.getInstance().unregister(jmxServerBean);
            }
        } catch (Exception e) {
            LOG.warn("Failed to unregister with JMX", e);
        }
        jmxServerBean = null;
    }
    
    @Override
    public String getState() {
        return "leader";
    }

    /**
     * Returns the id of the associated QuorumPeer, which will do for a unique
     * id of this server. 
     */
    @Override
    public long getServerId() {
        return self.getId();
    }    
    
    @Override
    protected void revalidateSession(ServerCnxn cnxn, long sessionId,
        int sessionTimeout) throws IOException, InterruptedException {
        super.revalidateSession(cnxn, sessionId, sessionTimeout);
        try {
            // setowner as the leader itself, unless updated
            // via the follower handlers
            setOwner(sessionId, ServerCnxn.me);
        } catch (SessionExpiredException e) {
            // this is ok, it just means that the session revalidation failed.
        }
    }
}

File: zookeeper/server/quorum/LeaderZooKeeperServer.cpp
/*
 * LeaderZooKeeperServer.cpp
 *
 *  Created on: 2017-11-22
 *      Author: cxxjava@163.com
 */

#include "./Leader.hh"
#include "./QuorumPeer.hh"
#include "./LeaderZooKeeperServer.hh"
#include "./ProposalRequestProcessor.hh"
#include "./ToBeAppliedRequestProcessor.hh"
#include "../PrepRequestProcessor.hh"

namespace efc {
namespace ezk {

LeaderZooKeeperServer::LeaderZooKeeperServer(sp<FileTxnSnapLog> logFactory, QuorumPeer* self,
		sp<ZKDatabase> zkDb) THROWS(EIOException) :
		QuorumZooKeeperServer(logFactory, self->tickTime, self->minSessionTimeout,
				self->maxSessionTimeout, zkDb, self) {
}

sp<Leader> LeaderZooKeeperServer::getLeader() {
	return self->leader;
}

int LeaderZooKeeperServer::getGlobalOutstandingLimit() {
	return ZooKeeperServer::getGlobalOutstandingLimit() / (self->getQuorumSize() - 1);
}

void LeaderZooKeeperServer::createSessionTracker() {
	sessionTracker = new SessionTrackerImpl(shared_from_this(), getZKDatabase()
			->getSessionWithTimeOuts(), tickTime, self->getId(),
			getZooKeeperServerListener());
}

void LeaderZooKeeperServer::startSessionTracker() {
	sp<SessionTrackerImpl> st = dynamic_pointer_cast<SessionTrackerImpl>(sessionTracker);
	EThread::setDaemon(st, true); //!!!
	st->start();
}

void LeaderZooKeeperServer::setupRequestProcessors() {
	sp<RequestProcessor> finalProcessor = new FinalRequestProcessor(shared_from_this());
	sp<RequestProcessor> toBeAppliedProcessor = new ToBeAppliedRequestProcessor(
			finalProcessor, getLeader()->toBeApplied);
	commitProcessor = new CommitProcessor(toBeAppliedProcessor,
			ELLong::toString(getServerId()), false,
			getZooKeeperServerListener());
	EThread::setDaemon(commitProcessor, true); //!!!
	commitProcessor->start();
	sp<ProposalRequestProcessor> proposalProcessor = new ProposalRequestProcessor(dynamic_pointer_cast<LeaderZooKeeperServer>(shared_from_this()),
			commitProcessor);
	proposalProcessor->initialize();
	sp<PrepRequestProcessor> prp = new PrepRequestProcessor(shared_from_this(), proposalProcessor);
    firstProcessor = prp;
    EThread::setDaemon(prp, true); //!!!
    prp->start();
}

llong LeaderZooKeeperServer::getServerId() {
	return self->getId();
}

void LeaderZooKeeperServer::revalidateSession(sp<ServerCnxn> cnxn, llong sessionId,
	int sessionTimeout) THROWS(EIOException) {
	ZooKeeperServer::revalidateSession(cnxn, sessionId, sessionTimeout);
	try {
		// setowner as the leader itself, unless updated
		// via the follower handlers
		setOwner(sessionId, ServerCnxn::me);
	} catch (SessionExpiredException& e) {
		// this is ok, it just means that the session revalidation failed.
	}
}

} /* namespace ezk */
} /* namespace efc */
